File: lab04/lab04.rs
use rand::Rng;

fn main()
{
	// Настройка генераторов
	let mut rng_thread = rand::thread_rng();
	let mut lcg = LCG::new(12345, 1103515245, 12345, 1664525);
	
	struct LCG
	{
		state: u32,
		a: u32,
		c: u32,
		m: u32,
	}
	
	impl LCG
	{
		fn new(seed: u32, a: u32, c: u32, m: u32) -> Self
		{
			LCG
			{
				state: seed, a, c, m  
			}
		}
	
		fn next(&mut self) -> u32
		{
			self.state = (self.a * self.state + self.c) % self.m;
			self.state
		}
	}
	
	// Функция для вычисления выборочного среднего и дисперсии
	fn calculate_stats(data: &[f64]) -> (f64, f64)
	{
		let mean = data.iter().sum::<f64>() / data.len() as f64;
		let variance = data.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / data.len() as f64;
		(mean, variance)
	}
	
	// Генерация выборок
	let sample_size = 100000;
	let mut thread_rng_samples = vec![0.0; sample_size];
	let mut lcg_samples = vec![0.0; sample_size];
	
	for i in 0..sample_size
	{
		thread_rng_samples[i] = rng_thread.gen::<f64>();
		lcg_samples[i] = lcg.next() as f64 / lcg.m as f64;
	}
	
	let (thread_mean, thread_variance) = calculate_stats(&thread_rng_samples);
	let (lcg_mean, lcg_variance) = calculate_stats(&lcg_samples);
	
	println!("Выборочное среднее (thread_rng): {}", thread_mean);
	println!("Выборочная дисперсия (thread_rng): {}", thread_variance);
	println!("Выборочное среднее (LCG): {}", lcg_mean);
	println!("Выборочная дисперсия (LCG): {}", lcg_variance);
}

File: lab04/lab04.r
sample_size <- 100000

# Мультипликативный конгруэнтный генератор
a <- 16807 # Классический множитель для МКГ
m <- 2^31 - 1 # 2147483647; большое простое число
seed <- 1
lcg_values <- numeric(sample_size) # Вектор для значений
current <- seed
for (i in 1:sample_size)
{
  current <- (a * current) %% m # Остаток от деления
  lcg_values[i] <- current / m  # Нормализация в [0, 1)
}

# Встроенный генератор R
set.seed(1)
r_values <- runif(sample_size, min = 0, max = 1) # Генерация из равномерного распределения

# Функции для вычисления среднего и дисперсии
mean <- function(x) sum(x) / length(x)
variance <- function(x) sum((x - mean(x))^2) / (length(x) - 1)

# Результаты для МКГ
lcg_mean <- mean(lcg_values)
lcg_variance <- variance(lcg_values)

cat("Мультипликативный конгруэнтный генератор\n")
cat(sprintf("Среднее: %.6f\n", lcg_mean))
cat(sprintf("Дисперсия: %.6f\n\n", lcg_variance))

# Результаты для встроенного генератора R
r_mean <- mean(r_values)
r_variance <- variance(r_values)

cat("Встроенный генератор R\n")
cat(sprintf("Среднее: %.6f\n", r_mean))
cat(sprintf("Дисперсия: %.6f\n\n", r_variance))
